doc/FAQ.md

```diff
@@ -35,3 +35,12 @@ example, you can wrap your calls to the API into a seperate process, then use
 interprocess communication to relay the information. I do not plan on adding
 anything to the API to do this automatically though.
 
+## I am receiving `UnauthorizedAccessException` when attaching to a Linux process
+
+You need `PTRACE` capabilities (see `CAP_SYS_PTRACE`) and various associated
```

Comment:
```suggestion
You need `ptrace` capabilities to attach to the target process. One option to pass the access checks is running the inspecting process as root. If however you are just trying to inspect your own process, you may explicitly grant access to the inspecting process by performing a P/Invoke call to `prctl` with the `PR_SET_PTRACER` option.
```

Starting the inspecting process as a child is a reasonable (and recommended) way to exchange process ids, but I feel it should go into tutorials/samples instead.

Reply:
Hmm, not too happy with the suggested rewording

> [...] If however you are just trying to inspect your own process [...]

I mentioned "inspecting your own process" as a reference to the first FAQ entry, reducing that and removing the advice to start a child process makes this part look a bit out of place (I wouldn't understand what it means in the context of the rest of the reworded text).

As far as examples go, I thought about it but couldn't see a straightforward way to turn my test code into a useful example, sorry. I'm not used to code multiple processes in .NET Core.

Reply:
Oh sorry I forgot the first FAQ entry. I'll have a look into this again.

Reply:
```markdown
You need `ptrace` capabilities to attach to the target process. The followings are different ways to grant access to the inspecting process:

- Run the inspecting process as root.
- Run the target process from the inspecting process.
- If you are trying to [inspect your own process](#can-i-use-this-api-to-inspect-my-own-process), you may perform a P/Invoke call to `prctl` with the inspecting process ID and the `PR_SET_PTRACER` option.
```

After some consideration I think that advice (starting a new inspecting process ~as a child~) is more suitable for the first section, so it looks like:

```markdown
You may...

- `CreateSnapshotAndAttach` on Windows.
- Fork and `SuspendAndAttachToProcess` then kill the fork on Linux.
- Start a new inspecting process and `SuspendAndAttachToProcess`.
```

I'm concerned about the difference between a **child** process and a **new** process. As I mentioned above a **child** process is sometimes treated in a special way (automatically granting access to its parent, etc.). I'm not sure which you are referring to exactly.

Reply:
I don't mind splitting the information, I've pushed an update.

> As I mentioned above a child process is sometimes treated in a special way (automatically granting access to its parent, etc.). I'm not sure which you are referring to exactly.

I don't know either, I only tested child processes since anything else would be too much work to test for me right now (have to get the PID into a running process). I don't mind getting rid of the term itself.

>  Fork and `SuspendAndAttachToProcess` then kill the fork on Linux.

Is that actually supported though? I don't know how `fork` works internally but I'd assume other threads continue running? That could be a problem for the .NET runtime if it doesn't know that it was forked and keeps running. The discussion on [another issue](https://github.com/microsoft/clrmd/issues/303#issuecomment-556508640) indicates its unclear on whether this will work, so I'm not sure if it should be included in the FAQ in this form.

> Run the target process from the inspecting process.

I don't know if thats enough to grant you ptrace access though, I didn't test it but the manpages sound like you need to do some `PTRACE_TRACEME` call after the `fork` but before the `execve`.

I don't mind leaving either of these in, just thought it would make sense to fact-check them first.

Reply:
> Fork and `SuspendAndAttachToProcess` then kill the fork on Linux.

This needs further evaluation. I was just giving an example.

> Run the target process from the inspecting process.

This is tested. I'll add tests for it later so don't worry.